### DISCUSSION TITLE
An algorithm of publish link generation

### DISCUSSION THREAD

**Discussion by bash-spbu** [jgraph/drawio#2604] [Q&A] [answered]
Hello!

Could you please clarify a bit what is an algorithm of drawio publishing link (`File | Publish | Link ...`) generation? 
Specifically, I am interested in the case when the whole diagram is encoded in the link and external data storages (like Google Drive) are not used.

I would imagine it is something like `XML -> UTF-8 -> Brotli compression -> Base64 Encode -> Embed to the full URL -> URL Encode for remaining parts`.
But feels like I missed something.
P.S. Unfortunately, I wasn't able to use your neat [tool](https://jgraph.github.io/drawio-tools/tools/convert.html) for such a URL decoding.

The background:

I am the JetBrains IDEA diagrams developer. We would like to make an integration with drawio so that every diagram generated by our products could be opened in your brilliant editor.

I already implemented an XML file generation and now I would like to make a button `Open in JGraph online editor` so that the diagram could be opened in your online editor by a single click.

Thank you in advance for your help and all the hard work on the drawio and related products. I personally find them one of the best known and use it a lot when needed.

Best wishes,
JetBrains team.

**Comment by bash-spbu** [comment, marked as answer]
Oh, reading carefully your [FAQ docs](https://www.diagrams.net/doc/faq/supported-location-hash-properties) I noticed the prefix letter `R` which denotes `raw` diagram data. Removing it revealed that the algorithm for compression is `XML -> URL Encode -> Deflate -> URL Encode -> Embed to the full URL with trailing #R`, so that now I can generate such link as well.

I am sorry for my inattentiveness.

**Comment by davidjgraph** [comment]
Hi,

Strictly it's URIencode, rather than URL encode, there is a tiny difference, but if you hit it you'll not be able to decode.

See:

https://jgraph.github.io/drawio-tools/tools/convert.html
https://github.com/jgraph/drawio-tools/blob/master/tools/convert.html

Encode is https://github.com/jgraph/drawio-tools/blob/master/tools/convert.html#L11
encodeURIComponent -> deflateRaw -> btoa (base64 encode).

And opposites for decode.

**Reply by bash-spbu** [reply]
Thank you for a detailed answer! 
I will study the provided implementation.
